# Мобильное приложение для лотерейной зависимости


**Цель:** разработать мобильное приложение, которое призвано помочь "лотерейным наркоманам" лучше оценить свои шансы на выигрыш. В данном проекте будет разработано логическое ядро приложения и произведен расчет вероятностей

Для первой версии приложения необходимо, чтобы мы сосредоточились на лотерее 6/49 и построили функции, которые позволяют пользователям отвечать на такие вопросы, как:

1. Какова вероятность выиграть Большой приз с одним билетом?
2. Какова вероятность выиграть Большой приз, если мы разыграем 40 разных билетов (или любое другое число)?
3. Какова вероятность наличия хотя бы пяти (или четырех, или трех, или двух) выигрышных номеров на одном билете?

**Данные** За основу для расчетов возьмем исторические [данные](https://www.kaggle.com/datascienceai/lottery-dataset), полученные от национальной лотереи 6/49 в Канаде. В наборе данных есть данные для 3 665 рисунков, датируемых с 1982 по 2018 год.

In [48]:
import pandas as pd

In [36]:
def factorial(n):
    return 1 if (n < 1) else n * factorial(n-1)

                                                     Сочетания без повторений
\begin{equation}
_nC_k = {n \choose k} =  \frac{n!}{k!(n-k)!}
\end{equation}

In [38]:
def combination (n,k):
    """Функция для вычисления сочетаний без повторений"""
    return factorial(n)/factorial(n-k)/factorial(k)    
print(combination(49,6))

13983816.0


## Вероятность выиграть приз для одного билета

Нужно построить функцию, которая вычисляет вероятность выигрыша Большого приза для любого заданного билета. Для каждого розыгрыша, шесть номеров берутся из 49 номеров, и игроки выигрывают большой приз, если шесть номеров на их билетах соответствуют всем шести выбранным номерам.

Пусть существуют следующие требования к функции:

1. Внутри приложения пользователь вводит шесть различных чисел от 1 до 49.
2. Под капотом, **шесть чисел будут приходить в виде списка Python** и служить в качестве входных данных для нашей функции.
3. Инженерная команда хочет, чтобы функция печатала значение вероятности дружелюбным способом — таким образом, чтобы люди без какой-либо вероятностной подготовки могли понять.

Ниже гапишу функцию one_ticket_probability (), которая принимает на вход список из шести уникальных чисел и выводит вероятность выигрыша понятным образом.

In [119]:
def one_ticket_probability(numbers):
    """вычисляет вероятность выигрыша Большого приза с одним билетом"""
    m=1/combination(49,6)
    return print("Ваши шансы выиграть Большой приз с номерами {} равны {}%. Другими словами, у вас есть 1 в {:,} шансы на победу.".format(numbers,m*100,combination(49,6)))

one_ticket_probability([1,2,3,4,5,6])
one_ticket_probability([22,42,36,14,35,16])

Ваши шансы выиграть Большой приз с номерами [1, 2, 3, 4, 5, 6] равны 7.151123842018516e-06%. Другими словами, у вас есть 1 в 13,983,816.0 шансы на победу.
Ваши шансы выиграть Большой приз с номерами [22, 42, 36, 14, 35, 16] равны 7.151123842018516e-06%. Другими словами, у вас есть 1 в 13,983,816.0 шансы на победу.


In [120]:
data=pd.read_csv("649.csv")

In [121]:
data.shape

(3665, 11)

In [63]:
data.head(3)

PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
0      649            1                0  6/12/1982               3   
1      649            2                0  6/19/1982               8   
2      649            3                0  6/26/1982               1   

   NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
0              11              12              14              41   
1              33              36              37              39   
2               6              23              24              27   

   NUMBER DRAWN 6  BONUS NUMBER  
0              43            13  
1              41             9  
2              39            34

In [64]:
data.tail(3)

PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
3662      649         3589                0  6/13/2018               6   
3663      649         3590                0  6/16/2018               2   
3664      649         3591                0  6/20/2018              14   

      NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
3662              22              24              31              32   
3663              15              21              31              38   
3664              24              31              35              37   

      NUMBER DRAWN 6  BONUS NUMBER  
3662              34            16  
3663              49             8  
3664              48            17

In [57]:
data[data["DRAW NUMBER"]==2965]

PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
3009      649         2965                0  6/20/2012               1   
3010      649         2965                1  6/20/2012               2   
3011      649         2965                2  6/20/2012               4   
3012      649         2965                3  6/20/2012               8   

      NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
3009               6              10              26              31   
3010               3              15              17              25   
3011              18              27              28              31   
3012              13              17              32              44   

      NUMBER DRAWN 6  BONUS NUMBER  
3009              38            11  
3010              42             0  
3011              33             0  
3012              48             0

## Сверка с историческими данными

Следующие требования к логике ядра приложения:

1. Внутри приложения пользователь вводит шесть различных чисел от 1 до 49.
2. Под капотом, шесть чисел будут приходить в виде списка Python и служить в качестве входных данных для нашей функции.
3. функция, которая печатает: число раз, когда выбранная комбинация встречалась в наборе данных Канады; и вероятность выигрыша Большого приза в следующем розыгрыше с этой комбинацией.

In [95]:
# def extract_numbers(row):
#     for label,i in data.iterrows():
#         a=set(i[['NUMBER DRAWN 1','NUMBER DRAWN 2','NUMBER DRAWN 3','NUMBER DRAWN 4','NUMBER DRAWN 5','NUMBER DRAWN 6']])
#         return a

def extract_numbers(row):
    """Функция для извлецения выигрышных номеров из исторического набор данных"""
    return set(row[4:10].values)

In [96]:
winning_num=data.apply(extract_numbers,axis=1)

In [159]:
def check_historical_occurence(user_num,winnig_set):
    '''a:проверяет, произошла ли определенная комбинация в наборе данных лотереи Канады
    user_num: a Python list
    winnig_set: a pandas Series'''
    a=set(user_num)==winnig_set
    number=a.sum()
    if number ==0:
        return print("В прошлом такой комбинации {} не было. Шанс на выигрыш {}%.Другими словами, у вас есть 1 к {:,} шансов на победу.".format(user_num,1/combination(49,6)*100,combination(49,6)))
    else:
        return print("Такая комбинация {} в прошлом встречалась {} раз. Шанс на выигрыш {}%.Другими словами, у вас есть 1 к {:,} шансов на победу.".format(user_num,number,1/combination(49,6)*100,combination(49,6)))

In [160]:
a=[33, 36, 37, 39, 8, 41]

In [161]:
check_historical_occurence(a,winning_num)

Такая комбинация [33, 36, 37, 39, 8, 41] в прошлом встречалась 1 раз. Шанс на выигрыш 7.151123842018516e-06%.Другими словами, у вас есть 1 к 13,983,816.0 шансов на победу.


In [162]:

test_input_4 = [3, 2, 44, 22, 1, 44]
check_historical_occurence(test_input_4,winning_num)

В прошлом такой комбинации [3, 2, 44, 22, 1, 44] не было. Шанс на выигрыш 7.151123842018516e-06%.Другими словами, у вас есть 1 к 13,983,816.0 шансов на победу.


## Вероятность для нескольких билетов

Тк зачастую игроки используют в розыгрыше больше 1го билета, напишу функцию,которая рассчитывает шанс на выигрыш для любого кол-ва биллетов

следующие требования к приложению:

1. Пользователь будет вводить количество различных билетов, которые они хотят играть (без ввода конкретных комбинаций, которые они намерены играть).
2. Наша функция будет видеть целое число от 1 до 13 983 816 (максимальное количество различных билетов).
3. Функция должна выводить информацию о вероятности выигрыша Большого приза в зависимости от количества разыгранных билетов.

In [166]:
def multi_ticket_probability(num):
    "Вычисляет вероятность Большого приза  для любого кол-ва билетов от 1 до 13983816 "
    m=num/combination(49,6)  
    return print('''Шанс выиграть большой приз с кол-м билетов {:,} {:.6f}%.
Другими словами есть 1 из {:,} шансов выиграть.'''.format(num, m*100,
                                                               int(1/m)))

In [164]:
b=[1, 10, 100, 10000, 1000000, 6991908, 13983816]

In [167]:
for i in b:
    multi_ticket_probability(i)

Шанс выиграть большой приз с кол-м билетов 1 0.000007%.
Другими словами есть 1 из 13,983,816 шансов выиграть.
Шанс выиграть большой приз с кол-м билетов 10 0.000072%.
Другими словами есть 1 из 1,398,381 шансов выиграть.
Шанс выиграть большой приз с кол-м билетов 100 0.000715%.
Другими словами есть 1 из 139,838 шансов выиграть.
Шанс выиграть большой приз с кол-м билетов 10,000 0.071511%.
Другими словами есть 1 из 1,398 шансов выиграть.
Шанс выиграть большой приз с кол-м билетов 1,000,000 7.151124%.
Другими словами есть 1 из 13 шансов выиграть.
Шанс выиграть большой приз с кол-м билетов 6,991,908 50.000000%.
Другими словами есть 1 из 2 шансов выиграть.
Шанс выиграть большой приз с кол-м билетов 13,983,816 100.000000%.
Другими словами есть 1 из 1 шансов выиграть.


## Вероятность для совпедения 5, 4, 3.. выигрышных чисел

В большинстве лотерей 6/49 есть призы, если билет игрока соответствует двум, трем, четырем или пяти из шести взятых чисел. Как следствие, пользователи могут быть заинтересованы в знании вероятности наличия двух, трех, четырех или пяти выигрышных номеров.

**Требования к логике приложения:**

Внутри приложения пользователь вводит данные:
шесть различных чисел от 1 до 49 и целое число от 2 до 5, представляющее количество ожидаемых выигрышных номеров
Функция выводит информацию о вероятности наличия введенного количества выигрышных номеров.

In [178]:
def probability_less_6(num):
    "Вычисляет вероятность получить 2.3,4,5 номеров из выигрышной комбинации "
    combination(49,num)
    sucess_outcomes=combination(6,num)*(49-num)
    p=sucess_outcomes/combination(49,6)
    return print("Шанс получить приз для {} номеров из всей комбинации {} %".format(num,round(p*100,4)))

In [179]:
l=[2,3,4,5]
for i in l:
    probability_less_6(i)

Шанс получить приз для 2 номеров из всей комбинации 0.005 %
Шанс получить приз для 3 номеров из всей комбинации 0.0066 %
Шанс получить приз для 4 номеров из всей комбинации 0.0048 %
Шанс получить приз для 5 номеров из всей комбинации 0.0019 %


## Заключение

В данном проекте разработано главным образом 4 функции:
one_ticket_probability() — вычисляет вероятность выигрыша для 1го билета
check_historical_occurrence() — проверяет содержится ли комбинации в историческом наборе данных
multi_ticket_probability() — вычисляет вероятность выйгрыша для любого кол-ва билетов от  1 and 13,983,816
probability_less_6() — вычисляет вероятность получить 2,3,4,5 выигрышных номеров